File: scripts/ollama_query.py
#!/usr/bin/env python
"""
Script para consultar a Ollama desde un proceso separado usando Python
Este script se ejecuta como un proceso independiente para evitar problemas de SIGTERM
"""

import sys
import os
import json
import subprocess
import time

def main():
    # Verificar argumentos
    if len(sys.argv) < 2:
        print("Error: Se requiere un prompt como primer argumento")
        sys.exit(1)
    
    # Configuración
    prompt = sys.argv[1]
    model = sys.argv[2] if len(sys.argv) > 2 else "llama3.1:8b"
    output_file = sys.argv[3] if len(sys.argv) > 3 else os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "temp", "ollama_response.txt")
    
    # Crear directorio de salida si no existe
    output_dir = os.path.dirname(output_file)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    
    # Mostrar información
    print(f"Consultando a Ollama (modelo: {model})...")
    print(f"Prompt: {prompt[:50]}{'...' if len(prompt) > 50 else ''}")
    
    try:
        # Ejecutar Ollama como un proceso separado
        result = subprocess.run(
            ["ollama", "run", model, prompt],
            capture_output=True,
            text=True,
            timeout=60
        )
        
        # Verificar si hubo errores
        if result.returncode != 0:
            print(f"Error al ejecutar Ollama: {result.stderr}")
            with open(output_file, "w", encoding="utf-8") as f:
                json.dump({
                    "success": False,
                    "error": "Error en la ejecución de Ollama",
                    "stderr": result.stderr
                }, f)
            sys.exit(1)
        
        # Guardar la respuesta en el archivo de salida
        with open(output_file, "w", encoding="utf-8") as f:
            json.dump({
                "success": True,
                "response": result.stdout.strip()
            }, f)
        
        print(f"Respuesta de Ollama guardada en: {output_file}")
        print(f"Respuesta (primeros 100 caracteres): {result.stdout[:100]}{'...' if len(result.stdout) > 100 else ''}")
        
    except subprocess.TimeoutExpired:
        print("Error: Tiempo de espera agotado al ejecutar Ollama")
        with open(output_file, "w", encoding="utf-8") as f:
            json.dump({
                "success": False,
                "error": "Tiempo de espera agotado"
            }, f)
        sys.exit(1)
    except Exception as e:
        print(f"Error inesperado: {str(e)}")
        with open(output_file, "w", encoding="utf-8") as f:
            json.dump({
                "success": False,
                "error": str(e)
            }, f)
        sys.exit(1)
    
    sys.exit(0)

if __name__ == "__main__":
    main()

File: scripts/ollama_query.js
/**
 * Script para consultar a Ollama desde un proceso separado
 * Este script se ejecuta como un proceso independiente para evitar problemas de SIGTERM
 */

const { exec } = require('child_process');
const fs = require('fs');
const path = require('path');

// Configuración
const OLLAMA_MODEL = process.argv[3] || 'llama3.1:8b';
const PROMPT = process.argv[2];
const OUTPUT_FILE = process.argv[4] || path.join(__dirname, '../temp/ollama_response.txt');

// Verificar que tenemos un prompt
if (!PROMPT) {
    console.error('Error: Se requiere un prompt como primer argumento');
    process.exit(1);
}

// Crear directorio de salida si no existe
const outputDir = path.dirname(OUTPUT_FILE);
if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
}

// Función para escapar caracteres especiales en el prompt
function escapePrompt(prompt) {
    return prompt.replace(/"/g, '\\"');
}

// Ejecutar Ollama como un proceso separado
console.log(`Consultando a Ollama (modelo: ${OLLAMA_MODEL})...`);
console.log(`Prompt: ${PROMPT.substring(0, 50)}${PROMPT.length > 50 ? '...' : ''}`);

// Comando para ejecutar Ollama
const comando = `ollama run ${OLLAMA_MODEL} "${escapePrompt(PROMPT)}"`;

// Ejecutar el comando
exec(comando, { timeout: 60000 }, (error, stdout, stderr) => {
    if (error) {
        console.error('Error al ejecutar Ollama:', error);
        fs.writeFileSync(OUTPUT_FILE, JSON.stringify({
            success: false,
            error: error.message,
            stderr: stderr
        }));
        process.exit(1);
    }
    
    // Guardar la respuesta en el archivo de salida
    fs.writeFileSync(OUTPUT_FILE, JSON.stringify({
        success: true,
        response: stdout.trim()
    }));
    
    console.log('Respuesta de Ollama guardada en:', OUTPUT_FILE);
    console.log(`Respuesta (primeros 100 caracteres): ${stdout.substring(0, 100)}${stdout.length > 100 ? '...' : ''}`);
    
    process.exit(0);
});
